src/Domain/Services/ExtendRestApi.php

```diff
@@ -0,0 +1,193 @@
+		if ( ! is_string( $namespace ) ) {
+			$this->throw_exception(
+				__( 'You must provide a plugin namespace when extending a Store REST endpoint.', 'woo-gutenberg-products-block' )
+			);
+		}
+
+		if ( ! is_string( $endpoint ) || ! in_array( $endpoint, $this->endpoints, true ) ) {
+			$this->throw_exception(
+				/* translators: 1: a list of endpoints. 2: endpoint provided by user. */
+				sprintf( __( 'You must provide a valid Store REST endpoint to extend, valid endpoints are: %1$s. You provided %2$s.', 'woo-gutenberg-products-block' ), implode( ', ', $this->endpoints ), $endpoint )
+			);
+		}
+
+		if ( ! is_callable( $schema_callback ) ) {
+			$this->throw_exception(
+				__( '$schema_callback must be a callable function.', 'woo-gutenberg-products-block' )
+			);
+		}
+
+		if ( ! is_callable( $data_callback ) ) {
+			$this->throw_exception(
+				__( '$data_callback must be a callable function.', 'woo-gutenberg-products-block' )
+			);
+		}
+
+		$this->extend_data[ $endpoint ][ $namespace ] = [
+			'schema_callback' => $schema_callback,
+			'data_callback'   => $data_callback,
+		];
+
+		return true;
+	}
+
+	/**
+	 * Returns the registered endpoint data
+	 *
+	 * @param string $endpoint    A valid identifier.
+	 * @param array  $passed_args Passed arguments from the Schema class.
+	 * @return array Returns an array with registered endpoint data.
+	 * @throws Exception If a registered callback throws an error, or silently logs it.
+	 */
+	public function get_endpoint_data( $endpoint, array $passed_args = [] ) {
+		$registered_data = (object) [];
+		if ( ! isset( $this->extend_data[ $endpoint ] ) ) {
+			return $registered_data;
+		}
+		foreach ( $this->extend_data[ $endpoint ] as $namespace => $callbacks ) {
+			$data = [];
+
+			try {
+				$data = $callbacks['data_callback']( ...$passed_args );
+			} catch ( Throwable $e ) {
+				$this->throw_exception( $e );
+				continue;
+			}
+
+			if ( ! is_array( $data ) ) {
+				$this->throw_exception( __( '$data_callback must return an array.', 'woo-gutenberg-products-block' ) );
+				continue;
+			}
+
+			$registered_data[ $namespace ] = $data;
+		}
+		return $registered_data;
+	}
+
+	/**
+	 * Returns the registered endpoint schema
+	 *
+	 * @param string $endpoint    A valid identifier.
+	 * @param array  $passed_args Passed arguments from the Schema class.
+	 * @return array Returns an array with registered schema data.
+	 * @throws Exception If a registered callback throws an error, or silently logs it.
+	 */
+	public function get_endpoint_schema( $endpoint, array $passed_args = [] ) {
+		$registered_schema = [];
+		if ( ! isset( $this->extend_data[ $endpoint ] ) ) {
+			return [];
+		}
+
+		foreach ( $this->extend_data[ $endpoint ] as $namespace => $callbacks ) {
+			$schema = [];
+
+			try {
+				$schema = $callbacks['schema_callback']( ...$passed_args );
+			} catch ( Throwable $e ) {
+				$this->throw_exception( $e );
+				continue;
+			}
+
+			if ( ! is_array( $schema ) ) {
```

Comment:
I'm not sure if this will be cleaner so feel free to ignore, but you could do the array check within the try/catch block so throw_exception is called once in this method.